File: js/signoff_policy.js
// Copyright 2026 FireCal LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

'use strict';

const { SignoffPolicyType } = require('./enums');
const { OrganizationSnapshot } = require('./organization_snapshot');
const { PersonSnapshot } = require('./person_snapshot');

class SignoffPolicy {
  constructor({
    id,
    type,
    allowedUsers = [],
    allowedOrgs = [],
    allowedRoles = [],
    completed = false,
    completionTimestamp = null,
    signoffRecord = null,
  }) {
    this.id = id;
    this.type = type;
    this.allowedUsers = Object.freeze([...allowedUsers]);
    this.allowedOrgs = Object.freeze([...allowedOrgs]);
    this.allowedRoles = Object.freeze([...allowedRoles]);
    this.completed = completed;
    this.completionTimestamp = completionTimestamp;
    this.signoffRecord = signoffRecord;
    Object.freeze(this);
  }

  /**
   * @param {string} userId
   * @param {string} taskbookOwnerId
   * @param {Object<string, Array<'admin'|'officer'|'member'>>} orgMemberships
   *   Map from organization canonical ID (matching
   *   OrganizationSnapshot.source.canonicalId) to the list of OrgRoles
   *   values the user holds in that org. Un-sourced snapshots in
   *   allowedOrgs cannot be globally identified and will not match.
   *   See "Matching organizations" in schemas/signoff_policy.md.
   * @returns {boolean}
   */
  isEligible(userId, taskbookOwnerId, orgMemberships) {
    switch (this.type) {
      case SignoffPolicyType.THIS_USER:
        return userId === taskbookOwnerId;
      case SignoffPolicyType.SPECIFY_USERS:
        return this.allowedUsers.includes(userId);
      case SignoffPolicyType.ORG_MEMBERS:
        for (const org of this.allowedOrgs) {
          const canonicalId = org && org.source ? org.source.canonicalId : null;
          if (!canonicalId) continue;
          const roles = orgMemberships ? orgMemberships[canonicalId] : null;
          if (!roles) continue;
          if (this.allowedRoles.length === 0) return true;
          if (this.allowedRoles.some((r) => roles.includes(r))) return true;
        }
        return false;
      default:
        return false;
    }
  }

  /**
   * Snapshot-based alternative to isEligible. Returns whether the
   * signer is eligible to sign this policy when both the signer and
   * the taskbook owner are available as PersonSnapshot values.
   *
   * Matching uses full canonical_id + canonical_source equality on
   * the relevant source fields. Un-sourced snapshots never match.
   * See "Matching organizations" in schemas/signoff_policy.md.
   *
   * @param {PersonSnapshot} signer
   * @param {PersonSnapshot} taskbookOwner
   * @returns {boolean}
   */
  isEligibleFor(signer, taskbookOwner) {
    const sSource = signer && signer.source ? signer.source : null;
    const oSource = taskbookOwner && taskbookOwner.source ? taskbookOwner.source : null;

    switch (this.type) {
      case SignoffPolicyType.THIS_USER:
        if (!sSource || !oSource) return false;
        if (!sSource.canonicalId || !oSource.canonicalId) return false;
        if (!sSource.canonicalSource || !oSource.canonicalSource) return false;
        return sSource.canonicalId === oSource.canonicalId &&
               sSource.canonicalSource === oSource.canonicalSource;

      case SignoffPolicyType.SPECIFY_USERS: {
        const id = sSource ? sSource.canonicalId : null;
        if (!id) return false;
        return this.allowedUsers.includes(id);
      }

      case SignoffPolicyType.ORG_MEMBERS: {
        const memberships = (signer && signer.memberships) || [];
        for (const allowed of this.allowedOrgs) {
          const a = allowed && allowed.source ? allowed.source : null;
          if (!a || !a.canonicalId || !a.canonicalSource) continue;
          for (const membership of memberships) {
            const org = membership ? membership.organization : null;
            const m = org && org.source ? org.source : null;
            if (!m || !m.canonicalId || !m.canonicalSource) continue;
            if (a.canonicalId === m.canonicalId &&
                a.canonicalSource === m.canonicalSource) {
              if (this.allowedRoles.length === 0) return true;
              if (this.allowedRoles.some((r) => membership.roles.includes(r))) return true;
            }
          }
        }
        return false;
      }

      default:
        return false;
    }
  }

  /**
   * Returns true iff the policy is in a compliant state per the signing
   * contract in schemas/signoff_policy.md.
   *
   * - An unsigned policy (completed = false) is always compliant.
   * - A signed policy is compliant iff completionTimestamp and
   *   signoffRecord are both set and signoffRecord.signedAt equals
   *   completionTimestamp.
   * @returns {boolean}
   */
  isValidSigned() {
    if (!this.completed) return true;
    if (this.completionTimestamp == null) return false;
    if (this.signoffRecord == null) return false;
    const a = this.signoffRecord.signedAt;
    const b = this.completionTimestamp;
    if (a === b) return true;
    if (a instanceof Date && b instanceof Date) return a.getTime() === b.getTime();
    return false;
  }
}

/**
 * Returns true if a policy list is satisfied given a require-all flag.
 * Empty policy list is always OK.
 */
function signoffsOK(policies, requireAll) {
  if (!policies || policies.length === 0) return true;
  return requireAll
    ? policies.every((p) => p.completed)
    : policies.some((p) => p.completed);
}

module.exports = { SignoffPolicy, signoffsOK };

File: dart/lib/signoff_policy.dart
// Copyright 2026 FireCal LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import 'enums.dart';
import 'organization_snapshot.dart';
import 'person_snapshot.dart';
import 'signoff_record.dart';

/// Defines who may sign off on a taskbook, section, or task.
class SignoffPolicy {
  final String id;
  final SignoffPolicyType type;
  final List<String> allowedUsers;
  final List<OrganizationSnapshot> allowedOrgs;
  final List<OrgRoles> allowedRoles;
  final bool completed;
  final DateTime? completionTimestamp;
  final SignoffRecord? signoffRecord;

  const SignoffPolicy({
    required this.id,
    required this.type,
    this.allowedUsers = const [],
    this.allowedOrgs = const [],
    this.allowedRoles = const [],
    this.completed = false,
    this.completionTimestamp,
    this.signoffRecord,
  });

  /// Returns whether [userId] is eligible to sign this policy.
  ///
  /// [taskbookOwnerId] is used only when [type] is [SignoffPolicyType.thisUser].
  ///
  /// [orgMemberships] maps organization canonical ID (matching
  /// [OrganizationSnapshot.source.canonicalId]) to the list of roles
  /// [userId] holds in that organization. Un-sourced snapshots in
  /// [allowedOrgs] cannot be globally identified and will not match.
  /// See "Matching organizations" in schemas/signoff_policy.md.
  bool isEligible(
    String userId,
    String taskbookOwnerId,
    Map<String, List<OrgRoles>> orgMemberships,
  ) {
    switch (type) {
      case SignoffPolicyType.thisUser:
        return userId == taskbookOwnerId;

      case SignoffPolicyType.specifyUsers:
        return allowedUsers.contains(userId);

      case SignoffPolicyType.orgMembers:
        for (final org in allowedOrgs) {
          final canonicalId = org.source?.canonicalId;
          if (canonicalId == null) continue;
          final roles = orgMemberships[canonicalId];
          if (roles == null) continue;
          if (allowedRoles.isEmpty) return true;
          if (allowedRoles.any(roles.contains)) return true;
        }
        return false;
    }
  }

  /// Snapshot-based alternative to [isEligible]. Returns whether
  /// [signer] is eligible to sign this policy when both the signer and
  /// the taskbook owner are available as [PersonSnapshot] values.
  ///
  /// Matching uses full `canonical_id + canonical_source` equality on
  /// the relevant `source` fields. Un-sourced snapshots never match.
  /// See "Matching organizations" in schemas/signoff_policy.md.
  bool isEligibleFor(PersonSnapshot signer, PersonSnapshot taskbookOwner) {
    switch (type) {
      case SignoffPolicyType.thisUser:
        final s = signer.source;
        final o = taskbookOwner.source;
        if (s == null || o == null) return false;
        if (s.canonicalId == null || o.canonicalId == null) return false;
        if (s.canonicalSource == null || o.canonicalSource == null) return false;
        return s.canonicalId == o.canonicalId &&
            s.canonicalSource == o.canonicalSource;

      case SignoffPolicyType.specifyUsers:
        final id = signer.source?.canonicalId;
        if (id == null) return false;
        return allowedUsers.contains(id);

      case SignoffPolicyType.orgMembers:
        final memberships = signer.memberships ?? const [];
        for (final allowed in allowedOrgs) {
          final a = allowed.source;
          if (a == null || a.canonicalId == null || a.canonicalSource == null) {
            continue;
          }
          for (final membership in memberships) {
            final m = membership.organization.source;
            if (m == null || m.canonicalId == null || m.canonicalSource == null) {
              continue;
            }
            if (a.canonicalId == m.canonicalId &&
                a.canonicalSource == m.canonicalSource) {
              if (allowedRoles.isEmpty) return true;
              if (allowedRoles.any(membership.roles.contains)) return true;
            }
          }
        }
        return false;
    }
  }

  /// Pure. Returns `true` iff the policy is in a compliant state per
  /// the signing contract in schemas/signoff_policy.md.
  ///
  /// - An unsigned policy (`completed = false`) is always compliant.
  /// - A signed policy is compliant iff `completionTimestamp` and
  ///   `signoffRecord` are both set and `signoffRecord.signedAt`
  ///   equals `completionTimestamp`.
  bool isValidSigned() {
    if (!completed) return true;
    if (completionTimestamp == null) return false;
    if (signoffRecord == null) return false;
    return signoffRecord!.signedAt == completionTimestamp;
  }

  SignoffPolicy copyWith({
    bool? completed,
    DateTime? completionTimestamp,
    SignoffRecord? signoffRecord,
  }) {
    return SignoffPolicy(
      id: id,
      type: type,
      allowedUsers: allowedUsers,
      allowedOrgs: allowedOrgs,
      allowedRoles: allowedRoles,
      completed: completed ?? this.completed,
      completionTimestamp: completionTimestamp ?? this.completionTimestamp,
      signoffRecord: signoffRecord ?? this.signoffRecord,
    );
  }
}

/// Returns `true` if [policies] is OK given a [requireAll] flag.
/// An empty policy list is always OK.
bool signoffsOK(List<SignoffPolicy> policies, bool requireAll) {
  if (policies.isEmpty) return true;
  return requireAll
      ? policies.every((p) => p.completed)
      : policies.any((p) => p.completed);
}
